### PR TITLE
proxy: initial userspace tcp proxy

### PR DESCRIPTION
**Pull request by xiang90**
/cc @heyitsanthony @gyuho 

prepare for etcdlet.

### CODE REVIEW

File: proxy/tcpproxy/userspace.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+func (tp *tcpProxy) run() error {
+	for {
+		in, err := tp.l.Accept()
+		if err != nil {
+			return err
+		}
+
+		go tp.serve(in)
```
Comment (gyuho) [inline proxy/tcpproxy/userspace.go]: Isn't this blocking? We don't need to wrap with go func( ?

Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: what do you mean? sorry. i do not follow.

File: proxy/tcpproxy/userspace.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	"time"
+)
+
+type tcpProxy struct {
+	l               net.Listener
+	monitorInterval time.Duration
+
+	mu      sync.Mutex // guards the following fields
+	remotes []*remote
+	next    int
```
Comment (heyitsanthony) [inline proxy/tcpproxy/userspace.go]: `nextRemote`?

File: proxy/tcpproxy/userspace.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+func (r *remote) inactivate() {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.inactive = true
+}
+
+func (r *remote) tryReactivate() {
+	conn, err := net.Dial("tcp", r.addr)
```
Comment (heyitsanthony) [inline proxy/tcpproxy/userspace.go]: what happens to `conn`?

Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: ah... we should close it...

Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: well... i think i close it at line 48. this is basically a tcp ping to know if the remote is reachable and should be marked as active.

File: proxy/tcpproxy/userspace.go (hunk 2 above)
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: sure.

File: proxy/tcpproxy/userspace.go (hunk 3 above)
Comment (heyitsanthony) [inline proxy/tcpproxy/userspace.go]: oh you're right... I misread `if err == nil` as an error path. Maybe `if err != nil { return }`, then the rest?

Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: oh. sure.

File: proxy/tcpproxy/userspace.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		}
+		remote.inactivate()
+	}
+
+	if out == nil {
+		in.Close()
+		return
+	}
+
+	go func() {
```
Comment (heyitsanthony) [inline proxy/tcpproxy/userspace.go]: just `go io.Copy(in, out)`?

File: proxy/tcpproxy/userspace.go (hunk 1 above)
Comment (gyuho) [inline proxy/tcpproxy/userspace.go]: @xiang90 Nvm. I thought of blocking goroutine like this:

```go
go wrap(forever())              // blocking
go func() { wrap(forever()) }() // non-blocking
```

but we don't wrap with anything, so don't matter.

File: proxy/tcpproxy/userspace_test.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		t.Fatal(err)
+	}
+	got, gerr := ioutil.ReadAll(res.Body)
+	res.Body.Close()
+	if gerr != nil {
+		t.Fatal(gerr)
+	}
+
+	if string(got) != want {
+		t.Errorf("got = %s, want %s", string(got), want)
```
Comment (gyuho) [inline proxy/tcpproxy/userspace_test.go]: I don't think we need `string(` because of `%s` here?

File: proxy/tcpproxy/userspace.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+	mu      sync.Mutex // guards the following fields
+	remotes []*remote
+	next    int
+}
+
+type remote struct {
+	mu       sync.Mutex
+	addr     string
+	inactive bool
```
Comment (heyitsanthony) [inline proxy/tcpproxy/userspace.go]: just `active`? negated variables are confusing (e.g., double-negatives like `if !inactive {}`)...

Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: the problem is the initial state should be active. but the default value of bool is false. so we need to be more careful about initializing remote. i do not have strong opinion on this.

Comment (heyitsanthony) [inline proxy/tcpproxy/userspace.go]: ok fair enough

File: proxy/tcpproxy/userspace.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+		in.Close()
+		out.Close()
+	}()
+
+	io.Copy(out, in)
+	out.Close()
+	in.Close()
+}
+
+// pick picks a active remote in round-robin fashion
```
Comment (heyitsanthony) [inline proxy/tcpproxy/userspace.go]: `pick picks a remote [...]`?

File: proxy/tcpproxy/userspace.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	}
+}
+
+func (r *remote) isActive() bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return !r.inactive
+}
+
+func (tp *tcpProxy) run() error {
```
Comment (heyitsanthony) [inline proxy/tcpproxy/userspace.go]: should this launch `runMonitor`?

Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: yes. i am working on it.

File: proxy/tcpproxy/userspace.go (hunk 4 above)
Comment (xiang90) [inline proxy/tcpproxy/userspace.go]: i do not think we want to leave a conn half open. so we close both side when either read/write meets an error?

**Comment by xiang90** [comment]
@gyuho @heyitsanthony All fixed. PTAL.

**Comment by gyuho** [comment]
LGTM. Thanks.

File: proxy/tcpproxy/userspace.go (hunk 4 above)
Comment (heyitsanthony) [inline proxy/tcpproxy/userspace.go]: OK, I see how it could block forever without this set of closes.

**Comment by heyitsanthony** [comment]
lgtm

**Pull request merged**